packages/smart-accounts/src/ma-v2/accounts/account.ts

```diff
@@ -0,0 +1,184 @@
+ * @returns {Promise<ModularAccountV2>} A MAv2 account.
+ */
+export async function toModularAccountV2<TMode extends Mode = Mode>({
+  client,
+  owner,
+  deferredAction,
+  signerEntity,
+  accountAddress: accountAddress_,
+  salt = 0n,
+  factoryAddress: factoryAddress_,
+  factoryData: factoryData_,
+  implementationAddress: implementationAddress_,
+  mode,
+}: ToModularAccountV2Params<TMode>): Promise<ModularAccountV2> {
+  const is7702 = mode === "7702";
+
+  const entityId = signerEntity?.entityId ?? DEFAULT_OWNER_ENTITY_ID;
+
+  const factoryAddress =
+    factoryAddress_ ??
+    (owner.type === "webAuthn"
+      ? DefaultAddress.MAV2_FACTORY_WEBAUTHN
+      : DefaultAddress.MAV2_FACTORY);
+
+  const implementationAddress =
+    implementationAddress_ ??
+    (is7702
+      ? DefaultAddress.SMAV2_7702
+      : owner.type === "webAuthn"
+        ? DefaultAddress.MAV2
+        : DefaultAddress.SMAV2_BYTECODE);
+
+  const getFactoryArgs = async () => {
+    if (is7702) {
+      return {
+        factory: "0x7702",
+        factoryData: "0x",
+      } as const;
+    }
+
+    if (owner.type === "webAuthn") {
+      const { x, y } = parsePublicKey(owner.publicKey);
+      return {
+        factory: factoryAddress,
+        factoryData:
+          factoryData_ ??
+          encodeFunctionData({
+            abi: accountFactoryAbi,
+            functionName: "createWebAuthnAccount",
+            args: [x, y, salt, entityId],
+          }),
+      };
+    }
+
+    return {
+      factory: factoryAddress,
+      factoryData:
+        factoryData_ ??
+        encodeFunctionData({
+          abi: accountFactoryAbi,
+          functionName: "createSemiModularAccount",
+          args: [owner.address, salt],
+        }),
+    };
+  };
+
+  const accountAddress =
+    accountAddress_ ??
+    predictModularAccountV2Address({
+      factoryAddress,
+      implementationAddress,
+      salt,
+      ...(owner.type === "webAuthn"
+        ? {
+            type: "WebAuthn",
+            ownerPublicKey: owner.publicKey,
+            entityId,
+          }
+        : {
+            type: "SMA", // TODO(jh): should this never be "MA"? double check w/ adam.
+            ownerAddress: owner.address,
+            entityId,
+          }),
+    });
+
+  let authorization: ToSmartAccountParameters["authorization"];
+  if (is7702) {
+    // TODO(v5): Ensure this works w/ our signer types.
+    if (owner.type !== "local") {
+      throw new InvalidOwnerError(
+        `Owner of type ${owner.type} is unsupported for 7702 mode.`
+      );
+    }
+    if (owner.signAuthorization == null) {
+      throw new InvalidOwnerError(
+        "Owner must implement `signAuthorization` to be used with 7702 mode."
+      );
+    }
+    if (
+      entityId === DEFAULT_OWNER_ENTITY_ID &&
+      owner.address !== accountAddress
+    ) {
+      throw new EntityIdOverrideError();
+    }
+    authorization = {
+      // The current version of Viem has some pretty strict constraints
```

Comment:
You mentioned this changed to be optional on `Account` on a later version of viem, right?

Reply:
This is what I was referring to: https://github.com/wevm/viem/blob/c88ae75376af7ed7cae920c25116804214a4fea3/src/accounts/types.ts#L114

`sign` & `signAuthorization` are optional on a `CustomSource` account now, but required on a `PrivateKeyAccount`. in v3, `sign` will be required on `CustomSource` accounts.